[JSC] Use `CachedCall` for iterator `next` method call

sosukesuzuki opened the pull request:
#### 5ee92514060caaeb7154dcedc8c0d10daab141ba
<pre>
[JSC] Use `CachedCall` for iterator `next` method call
<a href="https://bugs.webkit.org/show_bug.cgi?id=298184">https://bugs.webkit.org/show_bug.cgi?id=298184</a>

Reviewed by Yusuke Suzuki.

This patch changes to use `CachedCall` for iterator `next` method
call in `forEachInIterable` and `forEachInIterationRecord`.

                                              TipOfTree                  Patched

iterator-prototype-toArray-generator      366.0637+-10.7211    ^    296.6426+-3.9403        ^ definitely 1.2340x faster
uint8array-constructor-generator          315.6941+-1.3522     ^    254.6056+-2.4686        ^ definitely 1.2399x faster

Tests: JSTests/microbenchmarks/iterator-prototype-toArray-generator.js
       JSTests/microbenchmarks/uint8array-constructor-generator.js
       JSTests/stress/iterator-cached-call-bad-time.js
       JSTests/stress/iterator-cached-call-prototype-mutation.js

Tests: JSTests/microbenchmarks/iterator-prototype-toArray-generator.js
       JSTests/microbenchmarks/uint8array-constructor-generator.js
       JSTests/stress/iterator-cached-call-bad-time.js
       JSTests/stress/iterator-cached-call-prototype-mutation.js
* JSTests/microbenchmarks/iterator-prototype-toArray-generator.js: Added.
(shouldBe):
(gen):
* JSTests/microbenchmarks/uint8array-constructor-generator.js: Added.
(shouldBe):
(gen):
* JSTests/stress/iterator-cached-call-bad-time.js: Added.
(shouldBe):
(generator):
(iter.next):
* JSTests/stress/iterator-cached-call-prototype-mutation.js: Added.
(shouldBe):
(shouldThrow):
(throw.new.Error.generator):
(throw.new.Error.iter.forEach.):
(throw.new.Error):
(shouldBe.generator):
(shouldBe.iter.next):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/interpreter/CachedCall.cpp: Added.
(JSC::CachedCall::CachedCall):
(JSC::CachedCall::relink):
* Source/JavaScriptCore/interpreter/CachedCall.h:
(JSC::CachedCall::CachedCall): Deleted.
(JSC::CachedCall::relink): Deleted.
* Source/JavaScriptCore/interpreter/ProtoCallFrame.h:
* Source/JavaScriptCore/interpreter/ProtoCallFrameInlines.h:
* Source/JavaScriptCore/runtime/IteratorOperations.cpp:
(JSC::iteratorNextWithCachedCall):
(JSC::iteratorStepWithCachedCall):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInIterationRecord):
(JSC::forEachInIterable):
* Source/JavaScriptCore/wasm/WasmGlobal.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyGlobal.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyTable.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyWrapperFunction.h:

Canonical link: <a href="https://commits.webkit.org/300143@main">https://commits.webkit.org/300143@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2cc31a7239db5c5ac5903b9dc6136d1060457542

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41177 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31836 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127930 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73563 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41879 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49756 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92269 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124432 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33434 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108819 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72946 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32442 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26981 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71501 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/113610 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102920 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27155 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130755 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/120000 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48408 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36799 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100857 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48776 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105040 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100764 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25546 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46172 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24251 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/45104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48267 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53979 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150162 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47738 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/38324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51084 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49420 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.h:
```diff
@@ -153,12 +153,12 @@ using JSOrWasmInstruction = Variant<const JSInstruction*, uintptr_t /* IPIntOffs
         enum ExecutionFlag { Normal, InitializeAndReturn };
         
-        CodeBlock* prepareForCachedCall(CachedCall&, JSFunction*);
+        JS_EXPORT_PRIVATE CodeBlock* prepareForCachedCall(CachedCall&, JSFunction*);
```
This change is problematic since attaching this means we cannot do inlining even in PGO.

sosukesuzuki replied:
I made the following changes:

* Created a new `CachedCall.cpp`
* Moved the functions in `CachedCall` that call `prepareForCachedCall` (constructor and `relink`) to `CachedCall.cpp`
* Attached `JS_EXPORT_PRIVATE` to the `CachedCall` constructor

Is this acceptable?

sosukesuzuki commented on Source/JavaScriptCore/interpreter/CachedCall.cpp:
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2009-2023 Apple Inc. All rights reserved.
+ * Copyright (C) 2025 Codeblog CORP.
```
Legal name of bun

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -57,6 +60,28 @@ JSValue iteratorNext(JSGlobalObject* globalObject, IterationRecord iterationReco
+    ASSERT(JSC::getCallData(iterationRecord.nextMethod).type == CallData::Type::JS);
+
+    JSValue result;
+    if (!argument.isEmpty())
```
You can use `if (argument)`.

Constellation commented on Source/JavaScriptCore/runtime/IteratorOperations.cpp:
```diff
@@ -57,6 +60,28 @@ JSValue iteratorNext(JSGlobalObject* globalObject, IterationRecord iterationReco
+        result = cachedCall->callWithArguments(globalObject, iterator, argument);
+    else
+        result = cachedCall->callWithArguments(globalObject, iterator);
+    RETURN_IF_EXCEPTION(scope, JSValue());
```
We use `{ }` for new code instead of `JSValue()` here.

sosukesuzuki committed:
[JSC] Use `CachedCall` for iterator `next` method call
https://bugs.webkit.org/show_bug.cgi?id=298184

Reviewed by Yusuke Suzuki.

This patch changes to use `CachedCall` for iterator `next` method
call in `forEachInIterable` and `forEachInIterationRecord`.

                                              TipOfTree                  Patched

iterator-prototype-toArray-generator      366.0637+-10.7211    ^    296.6426+-3.9403        ^ definitely 1.2340x faster
uint8array-constructor-generator          315.6941+-1.3522     ^    254.6056+-2.4686        ^ definitely 1.2399x faster

Tests: JSTests/microbenchmarks/iterator-prototype-toArray-generator.js
       JSTests/microbenchmarks/uint8array-constructor-generator.js
       JSTests/stress/iterator-cached-call-bad-time.js
       JSTests/stress/iterator-cached-call-prototype-mutation.js

Tests: JSTests/microbenchmarks/iterator-prototype-toArray-generator.js
       JSTests/microbenchmarks/uint8array-constructor-generator.js
       JSTests/stress/iterator-cached-call-bad-time.js
       JSTests/stress/iterator-cached-call-prototype-mutation.js
* JSTests/microbenchmarks/iterator-prototype-toArray-generator.js: Added.
(shouldBe):
(gen):
* JSTests/microbenchmarks/uint8array-constructor-generator.js: Added.
(shouldBe):
(gen):
* JSTests/stress/iterator-cached-call-bad-time.js: Added.
(shouldBe):
(generator):
(iter.next):
* JSTests/stress/iterator-cached-call-prototype-mutation.js: Added.
(shouldBe):
(shouldThrow):
(throw.new.Error.generator):
(throw.new.Error.iter.forEach.):
(throw.new.Error):
(shouldBe.generator):
(shouldBe.iter.next):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/interpreter/CachedCall.cpp: Added.
(JSC::CachedCall::CachedCall):
(JSC::CachedCall::relink):
* Source/JavaScriptCore/interpreter/CachedCall.h:
(JSC::CachedCall::CachedCall): Deleted.
(JSC::CachedCall::relink): Deleted.
* Source/JavaScriptCore/interpreter/ProtoCallFrame.h:
* Source/JavaScriptCore/interpreter/ProtoCallFrameInlines.h:
* Source/JavaScriptCore/runtime/IteratorOperations.cpp:
(JSC::iteratorNextWithCachedCall):
(JSC::iteratorStepWithCachedCall):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInIterationRecord):
(JSC::forEachInIterable):
* Source/JavaScriptCore/wasm/WasmGlobal.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyGlobal.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyTable.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyWrapperFunction.h:

Canonical link: https://commits.webkit.org/300143@main
